### PR TITLE
Feature/fix bare citations script

### LINKED ISSUE #113: Feature: check for bare citation keys
Things like this: 
image

It should be possible to detect automatically, e.g. by compiling a list of citation keys in references.bib and checking to see if they occur in the markdown without a preceding "@".

### PR DESCRIPTION
**Pull request by cleong110**
Fixes #113 

Reads list of citation keys, uses regex to check for those keys, but if they don't have `@`, add it. 
Also, print out line number and key that was fixed.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,6 +18,7 @@ dst/thesis.pdf: dst/index_shortcode.md src/references.bib
 
 dst/index.md: src/index.md src/markdown_fix.sh src/formats.md dst tmp/datasets.md dst/assets
 	cat src/index.md > $@
+	python src/fix_bare_citations.py src/references.bib $@
```
Comment (AmitMY) [inline Makefile:21]: i think it would be best if this was not a "fixer" (since it does not fix the source code, and allows us to write errors)
instead, i think this should be a "linter" - write a small github action that runs this script, and throws an error if this happens.
then you get quick feedback in the PR

**Review by AmitMY — CHANGES_REQUESTED** [review]

**Comment by cleong110** [comment]
That would be better! I will set that up

File: Makefile (hunk 1 above)
Comment (cleong110) [reply Makefile:21]: Done!

**Comment by cleong110** [comment]
Updated the script, renamed it, added a deploy action that calls it. Should now give an error on PR if there is a bare citation. 

Also it no longer bugs you about citations in comments

File: src/find_bare_citations.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    # don't fix, just notify
+    # with open(markdown_file_path, "w") as file:
+    #     file.writelines(updated_content)
+
+    return issues
+
+
+if __name__ == "__main__":
+    if len(sys.argv) != 3:
```
Comment (AmitMY) [inline src/find_bare_citations.py:58]: always use `argparse` - so much more readable and convenient than sys.argv

File: src/find_bare_citations.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        print(f"Usage: {sys.argv[0]} <bib_file_path> <markdown_file_path>")
```
Comment (AmitMY) [inline src/find_bare_citations.py:59]: argparse will also take care of this

File: src/find_bare_citations.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            in_comment = True
+
+        if line.strip().startswith("```"):
+            in_code_block = not in_code_block
+
+        if not in_code_block and not in_comment:
+            for key in citation_keys:
+                pattern = re.compile(r"(?<!@)\b" + re.escape(key) + r"\b")
+                if pattern.search(line):
+                    issues.append(
```
Comment (AmitMY) [inline src/find_bare_citations.py:40]: use `yield` instead of appending to an array and returning the array. more economic that way

File: src/find_bare_citations.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+def find_bare_citations(markdown_file_path, citation_keys):
+    with open(markdown_file_path, "r") as file:
+        content = file.readlines()
+
+    # updated_content = []
+    issues = []
+    in_code_block = False
+    in_comment = False
+
+    for i, line in enumerate(content):
```
Comment (AmitMY) [inline src/find_bare_citations.py:28]: this is quite inefficient - for every line, for every citation.
a better approach could be:

1. remove all comments from content (`re.sub` of `<!--[\s\S]*-->` probably)
2. remove all code blocks from content (similar to above)

then just `for each citation` check if it is in content

it will make the code a lot cleaner (no need to check in code block / in citation)

File: src/find_bare_citations.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+# https://chatgpt.com/share/68cde216-5aeb-41e1-a4b2-93e0855f6b98
+import re
+import sys
+
+
+def extract_citation_keys(bib_file_path):
+    citation_keys = []
+    with open(bib_file_path, "r") as file:
+        for line in file:
+            match = re.match(r"@\w+\{([^,]+),", line)
```
Comment (AmitMY) [inline src/find_bare_citations.py:13]: you could run this match on the entire content, why line by line?

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: src/find_bare_citations.py (hunk 2 above)
Comment (cleong110) [reply src/find_bare_citations.py:58]: Good call! Will do.

File: src/find_bare_citations.py (hunk 6 above)
Comment (cleong110) [reply src/find_bare_citations.py:13]: Didn't think of it, tbh. Not familiar with regex, got ChatGPT to help me write it. 

Although, if we do the entire content could we still extract line numbers?

File: src/find_bare_citations.py (hunk 4 above)
Comment (cleong110) [reply src/find_bare_citations.py:40]: I will have to read up on how that works, I'll give it a go

File: src/find_bare_citations.py (hunk 5 above)
Comment (cleong110) [reply src/find_bare_citations.py:28]: My testing was entirely focused on "does it function", I didn't think about ways to make it more efficient.

File: src/find_bare_citations.py (hunk 6 above)
Comment (cleong110) [reply src/find_bare_citations.py:13]: The original conversation was at https://chatgpt.com/share/68cde216-5aeb-41e1-a4b2-93e0855f6b98.

Note to self: ChatGPT output seems to be more trouble than it's worth sometimes, requiring significant revision. Maybe just use it for the regex part next time

File: src/find_bare_citations.py (hunk 5 above)
Comment (cleong110) [reply src/find_bare_citations.py:30]: OK, wrote something to extend the citation key list by random strings and re-benchmark. 
```
python find_bare_citations.py ./references.bib ./index.md
Bibliography had 393 citations, checking for bare citations:
Bare-citation check complete after ~1.75 seconds
Bibliography had 493 citations, checking for bare citations:
Bare-citation check complete after ~2.24 seconds
Bibliography had 593 citations, checking for bare citations:
Bare-citation check complete after ~2.73 seconds
Bibliography had 693 citations, checking for bare citations:
Bare-citation check complete after ~3.16 seconds
Bibliography had 793 citations, checking for bare citations:
Bare-citation check complete after ~3.68 seconds
Bibliography had 893 citations, checking for bare citations:
Bare-citation check complete after ~4.18 seconds
Bibliography had 993 citations, checking for bare citations:
Bare-citation check complete after ~4.56 seconds
Bibliography had 1093 citations, checking for bare citations:
Bare-citation check complete after ~5.01 seconds
Bibliography had 1193 citations, checking for bare citations:
Bare-citation check complete after ~5.45 seconds
Bibliography had 1293 citations, checking for bare citations:
Bare-citation check complete after ~5.94 seconds
Bibliography had 1393 citations, checking for bare citations:
Bare-citation check complete after ~6.40 seconds
Bibliography had 1493 citations, checking for bare citations:
Bare-citation check complete after ~6.87 seconds
Bibliography had 1593 citations, checking for bare citations:
Bare-citation check complete after ~7.49 seconds
Bibliography had 1693 citations, checking for bare citations:
Bare-citation check complete after ~7.87 seconds
Bibliography had 1793 citations, checking for bare citations:
Bare-citation check complete after ~8.58 seconds
Bibliography had 1893 citations, checking for bare citations:
Bare-citation check complete after ~8.88 seconds
Bibliography had 1993 citations, checking for bare citations:
Bare-citation check complete after ~9.37 seconds
Bibliography had 2093 citations, checking for bare citations:
Bare-citation check complete after ~9.96 seconds
Bibliography had 2193 citations, checking for bare citations:
Bare-citation check complete after ~10.39 seconds
```

image

So here's the growth curve to beat. 

I suppose the other thing to check would be to add lines to index.md and see how that affects things, but I think this is good enough

Comment (cleong110) [reply src/find_bare_citations.py:30]: Rewrote it to:

1. strip out comments
2. strip out code blocks
3. Not split content into lines
4. find all matches of bare citation keys
5. **Now we cannot return line numbers,** so instead we return the whole line in which a bare citation is found

Benchmarking results using the same procedure as above:
```
python find_bare_citations.py ./references.bib ./index.md 
Bibliography had 393 citations, checking for bare citations:
Bare-citation-check complete after ~0.76 seconds
Bibliography had 493 citations, checking for bare citations:
Bare-citation-check complete after ~0.92 seconds
Bibliography had 593 citations, checking for bare citations:
Bare-citation-check complete after ~1.09 seconds
Bibliography had 693 citations, checking for bare citations:
Bare-citation-check complete after ~1.30 seconds
Bibliography had 793 citations, checking for bare citations:
Bare-citation-check complete after ~1.50 seconds
Bibliography had 893 citations, checking for bare citations:
Bare-citation-check complete after ~1.65 seconds
Bibliography had 993 citations, checking for bare citations:
Bare-citation-check complete after ~1.83 seconds
Bibliography had 1093 citations, checking for bare citations:
Bare-citation-check complete after ~1.99 seconds
Bibliography had 1193 citations, checking for bare citations:
Bare-citation-check complete after ~2.16 seconds
Bibliography had 1293 citations, checking for bare citations:
Bare-citation-check complete after ~2.35 seconds
Bibliography had 1393 citations, checking for bare citations:
Bare-citation-check complete after ~2.53 seconds
Bibliography had 1493 citations, checking for bare citations:
Bare-citation-check complete after ~2.72 seconds
Bibliography had 1593 citations, checking for bare citations:
Bare-citation-check complete after ~2.88 seconds
Bibliography had 1693 citations, checking for bare citations:
Bare-citation-check complete after ~3.20 seconds
Bibliography had 1793 citations, checking for bare citations:
Bare-citation-check complete after ~3.50 seconds
Bibliography had 1893 citations, checking for bare citations:
Bare-citation-check complete after ~3.55 seconds
Bibliography had 1993 citations, checking for bare citations:
Bare-citation-check complete after ~3.66 seconds
Bibliography had 2093 citations, checking for bare citations:
Bare-citation-check complete after ~3.87 seconds
Bibliography had 2193 citations, checking for bare citations:
Bare-citation-check complete after ~4.12 seconds
```

So ~~significantly~~ a lot faster when we get to high citation counts.

Comment (cleong110) [reply src/find_bare_citations.py]: [893fecf](https://github.com/sign-language-processing/sign-language-processing.github.io/pull/115/commits/893fecf199a50705ddeab154db46e8e3b13266bc) should resolve this one.

File: src/find_bare_citations.py (hunk 4 above)
Comment (cleong110) [reply src/find_bare_citations.py]: This change had a _dramatic_ speed-up effect on the function itself, as in the `find_bare_citations()` now takes 0.0 seconds. Iterating through the results does then take time. 

If I wish to output the lines  on which matches were found, the benchmarking results with `yield` look identical. 

```
python find_bare_citations.py ./references.bib ./index.md 
Bibliography had 393 citations, checking for bare citations:
Bare-citation check complete after ~0.74 seconds
Bibliography had 493 citations, checking for bare citations:
Bare-citation check complete after ~0.92 seconds
Bibliography had 593 citations, checking for bare citations:
Bare-citation check complete after ~1.09 seconds
Bibliography had 693 citations, checking for bare citations:
Bare-citation check complete after ~1.27 seconds
Bibliography had 793 citations, checking for bare citations:
Bare-citation check complete after ~1.49 seconds
Bibliography had 893 citations, checking for bare citations:
Bare-citation check complete after ~1.69 seconds
Bibliography had 993 citations, checking for bare citations:
Bare-citation check complete after ~1.83 seconds
Bibliography had 1093 citations, checking for bare citations:
Bare-citation check complete after ~2.00 seconds
Bibliography had 1193 citations, checking for bare citations:
Bare-citation check complete after ~2.17 seconds
Bibliography had 1293 citations, checking for bare citations:
Bare-citation check complete after ~2.36 seconds
Bibliography had 1393 citations, checking for bare citations:
Bare-citation check complete after ~2.69 seconds
Bibliography had 1493 citations, checking for bare citations:
Bare-citation check complete after ~2.82 seconds
Bibliography had 1593 citations, checking for bare citations:
Bare-citation check complete after ~2.97 seconds
Bibliography had 1693 citations, checking for bare citations:
Bare-citation check complete after ~3.19 seconds
Bibliography had 1793 citations, checking for bare citations:
Bare-citation check complete after ~3.63 seconds
Bibliography had 1893 citations, checking for bare citations:
Bare-citation check complete after ~3.54 seconds
Bibliography had 1993 citations, checking for bare citations:
Bare-citation check complete after ~3.71 seconds
Bibliography had 2093 citations, checking for bare citations:
Bare-citation check complete after ~3.91 seconds
Bibliography had 2193 citations, checking for bare citations:
Bare-citation check complete after ~4.18 seconds
Bibliography had 2293 citations, checking for bare citations:
Bare-citation check complete after ~4.32 seconds
```

Compare with https://github.com/sign-language-processing/sign-language-processing.github.io/pull/115#discussion_r1663016266, almost the same.

File: src/find_bare_citations.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import uuid
+
+
+def extract_citation_keys(bib_file_path):
+    citation_keys = []
+    with open(bib_file_path, "r") as file:
+        for line in file:
+            match = re.match(r"@\w+\{([^,]+),", line)
+            if match:
+                citation_keys.append(match.group(1))
```
Comment (AmitMY) [inline src/find_bare_citations.py:20]: seems like you didn't understand my note.

You can run `re. findall` on the ENTIRE file, and just extract all matches, without iterating line by line.
image

File: src/find_bare_citations.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+    # Remove HTML comments. regex from https://stackoverflow.com/a/28208465
+    content = re.sub("(<!--.*?-->)", "", content, flags=re.DOTALL)
+
+    # Remove Markdown code blocks, regex from https://stackoverflow.com/a/64116935
+    markdown_code_block_pattern = re.compile(r'```[^`]*```', re.DOTALL)
+
+    content = markdown_code_block_pattern.sub('', content)
+
+    for citation_key in citation_keys:
```
Comment (AmitMY) [inline src/find_bare_citations.py:36]: the faster way to doing this would be to first only search for the `citation_key` without a `@` (simple regex, single pass)
then, once finding it, extract the entire line using the match index.

right now, it will do lots of back-searches

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: src/find_bare_citations.py (hunk 7 above)
Comment (cleong110) [reply src/find_bare_citations.py:20]: No I understood, I don't know why it still has a "for line" in there. I'm pretty sure I took that out.

Comment (cleong110) [reply src/find_bare_citations.py:20]: Ah, dug into it: I thought this whole time we were trying to improve the regex for matching citations with index.md, not the regex for pulling citations out of references.bi

Comment (cleong110) [reply src/find_bare_citations.py:20]: Adding a "timeit" around the extraction-citations, I find that extracting the citations takes 0.0021669249999831663, is this worth fixing?

File: src/find_bare_citations.py (hunk 8 above)
Comment (cleong110) [reply src/find_bare_citations.py:36]: I don't understand, sorry. I am learning regexes for the first time here. Are we not doing exactly that? Line 39 is some kind of regex that searches the whole string of index.md, and then if there's a bare citation we go and extract the line in line using finditer. 

.search() only finds the first match, I wanted all matches, and the lines they are on.

I picked finditer because https://docs.python.org/3/library/re.html#re.finditer seemed to allow us to find all the matches.

findall would seem to be a possibility but the documentation isn't clear to me on how to get the full line, or what the return type is (I think just strings?), and also it wouldn't then work with `yield`-ing one bare-citation issue. 

I guess I could try to rewrite it to findall and then yield the list? If it does return a list? And then I don't know how the Match indexes come into this

Comment (cleong110) [reply src/find_bare_citations.py:36]: Tried switching to findall(), and eliminated search(). 

```
def find_bare_citations(markdown_file_path:Path, citation_keys:list)->list:
    
    content = markdown_file_path.read_text()

    # Remove HTML comments. regex from https://stackoverflow.com/a/28208465
    content = re.sub("(<!--.*?-->)", "", content, flags=re.DOTALL)

    # Remove Markdown code blocks, regex from https://stackoverflow.com/a/64116935
    markdown_code_block_pattern = re.compile(r'```[^`]*```', re.DOTALL)

    content = markdown_code_block_pattern.sub('', content)

    for citation_key in citation_keys:

        # magical regex from ChatGPT: captures the whole line that has a bare citation.
        pattern = re.compile(r'^.*(?<!@)(?:' + re.escape(citation_key) + r').*$', re.MULTILINE)

        # Find all matching lines
        matches = pattern.findall(content)
        if matches:
            issue_tuple = citation_key, matches
            yield issue_tuple
            
```
timing this, the time did go down, from 0.65 to 0.62 seconds. I'm guessing most of the issue is in the regex itself, which I do not know how to edit.

Comment (cleong110) [reply src/find_bare_citations.py:36]: Yup, if I add timeit calls around `matches = pattern.findall(content)`, it is taking up 0.60 of the 0.62 seconds. 

```
    regex_times = []

    for citation_key in citation_keys:

        # magical regex from ChatGPT: captures the whole line that has a bare citation.
        pattern = re.compile(r'^.*(?<!@)(?:' + re.escape(citation_key) + r').*$', re.MULTILINE)

        # Find all matching lines
        start_time = timeit.default_timer()
        matches = pattern.findall(content)
        regex_time = timeit.default_timer() -start_time
        regex_times.append(regex_time)
        if matches:
            issue_tuple = citation_key, matches
            yield issue_tuple

    print(f"Total time spent in findall calls: {sum(regex_times)}")
```

Comment (cleong110) [reply src/find_bare_citations.py:36]: OK, [had it explained to me another way](https://chatgpt.com/share/3c145c68-dc38-43f0-a2b1-27d3cd1d43f5), "simple regex, single pass" literally means 


```
# Find all positions of the citation key without the @ symbol
        key_pattern = re.compile(re.escape(citation_key))
```
As in, just find that exact string anywhere.

tl;dr, don't try to do EVERYTHING in the regex all at once. Finding the citation key is blazing fast with a regex, then finding newlines is not hard, and then seeing if there's an @ in front is a simple check.

File: src/find_bare_citations.py (hunk 7 above)
Comment (cleong110) [reply src/find_bare_citations.py]: Well, it wasn't that hard to fix once I figured out the other one. Editing this part to ALSO not go line-by-line, we save 

0.0021669249999831663 seconds - 0.0006211190002431977 seconds = 1.545806 milliseconds according to Google.

File: src/find_bare_citations.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            start_index = match.start()
+            line_start = content.rfind('\n', 0, start_index) + 1
+            line_end = content.find('\n', start_index)
+            if line_end == -1:
+                line_end = len(content)
+            line = content[line_start:line_end]
+
+            # Ensure the citation key is not immediately preceded by an @ symbol
+            if '@' not in content[start_index-1:start_index]:
+                matches.append(line)
```
Comment (AmitMY) [inline src/find_bare_citations.py:43]: after findging the `start_index`, this `if` should happen.
only then, if it did match, then you should look for `line_start`, `line_end`, etc
will not really change performance, but just better code

File: src/find_bare_citations.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+        if matches:
+            issue_tuple = citation_key, matches
+            yield issue_tuple
```
Comment (AmitMY) [inline src/find_bare_citations.py:47]:
```
yield citation_key, matches
```

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: src/find_bare_citations.py (hunk 10 above)
Comment (cleong110) [reply src/find_bare_citations.py:47]: That makes more sense

File: src/find_bare_citations.py (hunk 9 above)
Comment (cleong110) [reply src/find_bare_citations.py:43]: True, the rest of the code is redundant otherwise, might as well exit early.

**Comment by cleong110** [comment]
One thing I haven't been able to figure out is how to find the line number of an issue in the original content. 

I tried 
```
lines = content.splitlines()
```

followed later by 
```
line_index = lines.index(line)
```

Which adds almost zero overhead to run time. That was cool, I thought. 

But upon testing I found that if there's two identical lines for whatever reason, it will always find the first one.

**Comment by cleong110** [comment]
Another nice-to-have would be to show the index in the line, something like 

```
line 21, column 33:
    lorem ipsum foo_etal_2024 dolor sit amet
                            ^
```
And I got something working, but it failed in the case where we had a very long line. 


```
line 21, column 33:
    lorem ipsum foo_etal_2024 dolor sit amet blah blah
line wrap happens blah blah blah and now it looks bad
                            ^
```

**Review by AmitMY — APPROVED** [review]
